### DISCUSSION TITLE
Help ! how to show expression on export data on excel?

### DISCUSSION THREAD

**Discussion by sandhitya069** [serenity-is/Serenity#5500] [Help]
someone please tell me something about this one, i got error notice like this when i want to export data to excel with an expression inside. is it not possible to display an expressions when i export data to excel? 
![image](https://user-images.githubusercontent.com/55977575/108447228-7ba6ef80-729a-11eb-8e2c-9180ac497bce.png)

Thanks.

**Comment by luongnv6** [comment]
Hope help you

you need add MinSelectLevel(SelectLevel.List) attribute
```C#
 [DisplayName("City"), Expression("c.[Name]"), MinSelectLevel(SelectLevel.List)]
        public String CityName
        {
            get { return Fields.CityName[this]; }
            set { Fields.CityName[this] = value; }
        }

**Reply by sandhitya069** [reply]
it isn't work sir, i got notice like this sir
![image](https://user-images.githubusercontent.com/55977575/108451811-6f269500-72a2-11eb-930d-08ac29216439.png)

is it true that this is how to write?
![image](https://user-images.githubusercontent.com/55977575/108451923-9ed59d00-72a2-11eb-873a-9db1f416f2b0.png)

**Reply by luongnv6** [reply]
please show your code get Data to export

**Reply by sandhitya069** [reply]
here my code for PersonListReport
` [Report, RequiredPermission(PermissionKeys.General)]
    [Category("HRMS/Person"), DisplayName("Person List")]
    public class PersonListReport : IReport, IDataOnlyReport
    {
        protected ISqlConnections SqlConnections { get; }
        protected ITextLocalizer Localizer { get; }
        protected IServiceProvider ServiceProvider { get; }

        public PersonListReport(ISqlConnections sqlConnections,
            ITextLocalizer localizer, IServiceProvider serviceProvider)
        {
            SqlConnections = sqlConnections ?? throw new ArgumentNullException(nameof(sqlConnections));
            Localizer = localizer ?? throw new ArgumentNullException(nameof(localizer));
            ServiceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
        }

        public object GetData()
        {
            using var connection = SqlConnections.NewFor<Entities.PersonListRow>();
            var s = PersonListRow.Fields;
            return connection.List<PersonListRow>();
        }

        public List<ReportColumn> GetColumnList()
        {
            return ReportColumnConverter.ObjectTypeToList(typeof(Item), ServiceProvider, Localizer);
        }

        [BasedOnRow(typeof(HRMS.Entities.PersonListRow), CheckNames = true)]
        public class Item
        {
            public string PersonID { get; set; }
            public string FullName { get; set; }
            public string NickName { get; set; }
            public Gender PersonGender { get; set; }
            public MaritalStatus MaritalStatus { get; set; }
            public MasterStatus MasterStatus { get; set; }
            public string Address { get; set; }
            public string CityName { get; set; }
            public int PostalCode { get; set; }
            public int Phone { get; set; }
            public int Fax { get; set; }
        }
    }`

here my code for PersonListRow
`public sealed class PersonListRow : Row<PersonListRow.RowFields>, INameRow
    {   
        [DisplayName("Person Id"), Column("PersonID"), NotNull]
        public String PersonID
        {
            get => Fields.PersonID[this];
            set => Fields.PersonID[this] = value;
        }

        [DisplayName("Full Name"), Size(50), NotNull, QuickSearch, NameProperty]
        public String FullName
        {
            get => Fields.FullName[this];
            set => Fields.FullName[this] = value;
        }

        [DisplayName("Nick Name"), Column("NickName")]
        public String NickName
        {
            get => fields.NickName[this];
            set => fields.NickName[this] = value;
        }

        [DisplayName("Gender")]
        public Gender? PersonGender
        {
            get => Fields.PersonGender[this];
            set => Fields.PersonGender[this] = value;
        }

        [DisplayName("Status")]
        public MasterStatus? MasterStatus
        {
            get => Fields.MasterStatus[this];
            set => Fields.MasterStatus[this] = value;
        }

        [DisplayName("MRT Status")]
        public MaritalStatus? MaritalStatus
        {
            get => Fields.MaritalStatus[this];
            set => Fields.MaritalStatus[this] = value;
        }

        [DisplayName("Image")]
        [ImageUploadEditor(FilenameFormat = "PersonImage/~", CopyToHistory = true)]
        public String PersonImage
        {
            get => Fields.PersonImage[this];
            set => Fields.PersonImage[this] = value;
        }

        [DisplayName("Birthday")]
        public DateTime? Birthday
        {
            get => Fields.Birthday[this];
            set => Fields.Birthday[this] = value;
        }

        [DisplayName("Address"), Size(50), QuickSearch]
        public String Address
        {
            get => fields.Address[this];
            set => fields.Address[this] = value;
        }

        [DisplayName("City"), NotNull,
        ForeignKey("[dbo].[City]", "Oid"),
        LeftJoin("jCity"), TextualField("CityName")]
        [LookupEditor(typeof(CityRow),
        InplaceAdd = true),
        LookupInclude]
        public int? City
        {
            get { return Fields.City[this]; }
            set { Fields.City[this] = value; }
        }

        [DisplayName("City"), Expression("jCity.[CityName]"), MinSelectLevel(SelectLevel.List)]
        public String CityName
        {
            get { return Fields.CityName[this]; }
            set { Fields.CityName[this] = value; }
        }

        //[DisplayName("State")]
        //[Expression("jCity.[State]"), MinSelectLevel(SelectLevel.List)]
        //[ForeignKey("State", "Oid")]
        //[LeftJoin("jState")] //, TextualField("jState.[StateName]")]
        //[ReadOnly(true)]
        //public int? State
        //{
        //    get { return Fields.State[this]; }
        //    set { Fields.State[this] = value; }
        //}

        //[DisplayName("State"), Expression("jState.[StateName]"), MinSelectLevel(SelectLevel.List)]
        //[ReadOnly(true)]
        //public String StateName
        //{
        //    get { return Fields.StateName[this]; }
        //    set { Fields.StateName[this] = value; }
        //}

        //[DisplayName("Country")]
        //[Expression("jState.[Country]"), MinSelectLevel(SelectLevel.List)]
        //[ForeignKey("Countries", "Oid")]
        //[LeftJoin("jCountry")]//, TextualField("jCountry.[CountryName]")]
        //[ReadOnly(true)]
        //public int? Country
        //{
        //    get { return Fields.Country[this]; }
        //    set { Fields.Country[this] = value; }
        //}

        //[DisplayName("Country"), Expression("jCountry.[CountryName]"), MinSelectLevel(SelectLevel.List)]
        //[ReadOnly(true)]
        //public String CountryName
        //{
        //    get { return Fields.CountryName[this]; }
        //    set { Fields.CountryName[this] = value; }
        //}

        [DisplayName("Postal Code"), Size(50), QuickSearch]
        public String PostalCode
        {
            get => fields.PostalCode[this];
            set => fields.PostalCode[this] = value;
        }

        [DisplayName("Phone Number"), Size(50), QuickSearch]
        public String Phone
        {
            get => fields.Phone[this];
            set => fields.Phone[this] = value;
        }

        [DisplayName("Fax"), Size(50), QuickSearch]
        public String Fax
        {
            get => fields.Fax[this];
            set => fields.Fax[this] = value;
        }

        public PersonListRow()
        {
        }

        public PersonListRow(RowFields fields)
            : base(fields)
        {
        }

        public class RowFields : RowFieldsBase
        {
            public StringField PersonID;
            public StringField FullName;
            public StringField NickName;
            public StringField Address;
            public Int32Field City;
            public StringField CityName;
            //public Int32Field State;
            //public StringField StateName;
            //public Int32Field Country;
            //public StringField CountryName;
            public StringField PostalCode;
            public StringField Phone;
            public StringField Fax;
            public DateTimeField Birthday;
            public EnumField<MasterStatus> MasterStatus;
            public EnumField<MaritalStatus> MaritalStatus;
            public EnumField<Gender> PersonGender;
            public StringField PersonImage;

        }
    }`

**Reply by luongnv6** [reply]
Please change GetData
```C# 
public object GetData()
    {
        using var connection = SqlConnections.NewFor<Entities.PersonListRow>();
        var s = PersonListRow.Fields;

//select all column
        return connection.List<PersonListRow>(q=>
                q.SelectTableFields().SelectNonTableFields());
//or select a few columns 
       // return connection.List<PersonListRow>(q =>
               // q.Select(c.PersonID, c.FullName, c.CityName));
    }
